keps/sig-network/20190324-remove-kube-proxy-autocleanup.md

```diff
@@ -0,0 +1,74 @@
+---
+kep-number: 0
+title: Remove kube-proxy's automatic clean up logic
+authors:
+  - "@bowei"
+owning-sig: sig-network
+participating-sigs:
+reviewers:
+  - "@andrewsykim"
+approvers:
+  - "@bowei"
+  - "@thockin"
+editor:
+creation-date: 2018-03-24
+last-updated: 2018-03-24
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Remove kube-proxy's automatic clean up logic
+
+## Summary
+
+Remove the automatic network rule cleanup from kube-proxy.
+Only clean up rules on boot if the `--cleanup` flag is present.
+
+## Motivation
+
+kube-proxy's rule cleanup has a substantial performance impact,
+and is not normally necessary.
+
+### Goals
+
+* Reduce the typical kube-proxy startup time,
+by avoiding unnecessary cleanup.
+
+## Proposal
+
+Add the flag `--cleanup` to kube-proxy,
+and only clean up iptables/ipvs rules if `--cleanup` is set.
+If cleanup is not set,
+start kube-proxy without cleaning up old rules.
+
+`--cleanup` should run cleanup for both iptables and ipvs.
+There is overlap between modes and the kind of network rules used -
+for example,
+ipvs mode also creates some iptables rules.
+
+## Graduation Criteria
+
+Beta: Stable without notable bugs for at least one release.
```

Comment:
I'm not sure there's a good way to "beta" this - I predict nobody will really run the --cleanup flag.  I think this just goes straight to GA, with a strong release note about switching proxy modes requiring a reboot.

Reply:
I'm on the fence. If we added a --no-auto-cleanup feature argument, not many people would probably use it, but it would give us a huge opportunity to catch bugs caused by the change.